File: 0784-letter-case-permutation/0784-letter-case-permutation.py
class Solution:
    def letterCasePermutation(self, s: str) -> List[str]:
        result = []
        self.helper(0, s, '', result)
        return result
    
    def helper(self, i, str, input, output):
        # Base case
        if i == len(str):
            output.append(input)
            return
        
        if str[i].isalpha():
            # if it's a letter
            self.helper(i + 1, str, input + str[i].upper(), output)
            self.helper(i + 1, str, input + str[i].lower(), output)
        else:
            # if it's a digit    
            self.helper(i + 1, str, input + str[i], output)
        
        return

File: 0784-letter-case-permutation/0784-letter-case-permutation.js
/**
 * @param {string} s
 * @return {string[]}
 */
var letterCasePermutation = function(s) {
    const output = [];
    helper(0, s, '', output)
    return output;  
};

function helper(i, str, input, output) {
    // Base case
    if (i === str.length) {
        output.push(input);
        return;
    }
    
    if (isNaN(str[i])) {
        helper(i + 1, str, input + str[i].toLowerCase(), output);
        helper(i + 1, str, input + str[i].toUpperCase(), output);
    } else {
        helper(i + 1, str, input + str[i], output);
    }
    
    return;
}